File: backend/src/main/java/com/videodownloader/service/VideoProcessService.java
package com.videodownloader.service;

import org.springframework.stereotype.Service;
import java.io.*;
import java.nio.file.*;

@Service
public class VideoProcessService {

    // Dynamically resolve tools directory relative to the Java execution root
    private final String TOOLS_DIR = Paths.get(System.getProperty("user.dir"), "tools").toAbsolutePath().toString();
    private final String YT_DLP_PATH = Paths.get(TOOLS_DIR, "yt-dlp.exe").toString();

    public String getVideoInfo(String url) throws Exception {
        ProcessBuilder pb = new ProcessBuilder(
                YT_DLP_PATH,
                "-j",
                "--cookies-from-browser", "chrome",
                "--no-playlist",
                "--no-warnings",
                "--quiet",
                url);
        Process p = pb.start();

        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
        }
        int exitCode = p.waitFor();
        if (exitCode != 0) {
            throw new RuntimeException("Failed to extract video info: " + output);
        }
        return output.toString();
    }

    public File downloadVideo(String url, String formatId) throws Exception {
        File tempDir = Files.createTempDirectory("video-dl").toFile();
        File outputFile = new File(tempDir, "output.mp4");

        ProcessBuilder pb = new ProcessBuilder(
                YT_DLP_PATH,
                "-f", formatId,
                "--cookies-from-browser", "chrome",
                "--ffmpeg-location", TOOLS_DIR,
                "--merge-output-format", "mp4",
                "--concurrent-fragments", "10",
                "--http-chunk-size", "10M",
                "-o", outputFile.getAbsolutePath(),
                url);
        pb.redirectErrorStream(true);
        Process p = pb.start();

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                System.out.println("yt-dlp: " + line);
            }
        }

        int exitCode = p.waitFor();
        if (exitCode != 0) {
            throw new RuntimeException("Download process failed.");
        }

        return outputFile;
    }
}

File: dotnet_backend/Services/VideoProcessService.cs
using System.Diagnostics;
using System.Text;

namespace VideoDownloader.NET.Services;

public class VideoProcessService
{
    private readonly string _ytDlpPath;
    private readonly string _ffmpegPath;

    public VideoProcessService()
    {
        // Dynamically resolve the absolute path to the parent directory for tool access
        var projectDir = Directory.GetCurrentDirectory();
        var rootDir = Directory.GetParent(projectDir)?.FullName ?? projectDir;
        
        _ffmpegPath = Path.Combine(rootDir, "backend", "tools");
        _ytDlpPath = Path.Combine(_ffmpegPath, "yt-dlp.exe");
    }

    public async Task<string> GetVideoInfoAsync(string url)
    {
        var runArgs = new List<string> { "-j", "--no-playlist", "--no-warnings", "--quiet" };
        if (url.Contains("xhslink") || url.Contains("xiaohongshu"))
        {
            runArgs.Add("--cookies-from-browser");
            runArgs.Add("chrome");
        }
        runArgs.Add(url);

        var processInfo = new ProcessStartInfo
        {
            FileName = _ytDlpPath,
            Arguments = string.Join(" ", runArgs),
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };

        using var process = Process.Start(processInfo);
        if (process == null) throw new Exception("Failed to start yt-dlp process.");

        var output = await process.StandardOutput.ReadToEndAsync();
        var error = await process.StandardError.ReadToEndAsync();
        await process.WaitForExitAsync();

        if (process.ExitCode != 0)
        {
            throw new Exception($"Failed to extract video info: {error}");
        }

        return output;
    }

    public async Task<string> DownloadVideoAsync(string url, string formatId, string type = "video")
    {
        var tempDir = Path.Combine(Path.GetTempPath(), "video-dl-" + Guid.NewGuid());
        Directory.CreateDirectory(tempDir);
        var outputFile = Path.Combine(tempDir, type == "audio" ? "output.m4a" : "output.mp4");

        var runArgs = new List<string>
        {
            "-f", formatId,
            "--ffmpeg-location", _ffmpegPath,
            "--concurrent-fragments", "10",
            "--http-chunk-size", "10M"
        };

        if (type != "audio")
        {
            runArgs.Add("--merge-output-format");
            runArgs.Add("mp4");
        }
        
        if (url.Contains("xhslink") || url.Contains("xiaohongshu"))
        {
            runArgs.Add("--cookies-from-browser");
            runArgs.Add("chrome");
        }
        
        runArgs.Add("-o");
        runArgs.Add(outputFile);
        runArgs.Add(url);

        var processInfo = new ProcessStartInfo
        {
            FileName = _ytDlpPath,
            Arguments = string.Join(" ", runArgs),
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };

        using var process = Process.Start(processInfo);
        if (process == null) throw new Exception("Failed to start yt-dlp process.");

        process.OutputDataReceived += (s, e) => { if (e.Data != null) Console.WriteLine($"yt-dlp: {e.Data}"); };
        process.ErrorDataReceived += (s, e) => { if (e.Data != null) Console.WriteLine($"yt-dlp err: {e.Data}"); };
        
        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        await process.WaitForExitAsync();

        if (process.ExitCode != 0)
        {
            throw new Exception("Download process failed.");
        }

        return outputFile;
    }
}
